### PR TITLE
ISPN-7540 Calling getCache(String, String) - template overrides cache

### PR DESCRIPTION
**Pull request by wburns**
definition

* Throw exception when redefining a cache

https://issues.jboss.org/browse/ISPN-7540

### CODE REVIEW

**Comment by wburns** [comment]
Trying this again, tests should be in a lot better shape now.

**Comment by wburns** [comment]
Hoping this commit will fix remaining issues.

**Comment by wburns** [comment]
Previous run all passed http://ci.infinispan.org/viewLog.html?buildId=51449&tab=buildResultsDiv&buildTypeId=bt9&logTab= 

Running again to make sure with new changes and I can rebase.

**Comment by wburns** [comment]
Tests all seem good now as it ran again without failure http://ci.infinispan.org/viewLog.html?buildId=51460&buildTypeId=bt9

**Comment by wburns** [comment]
Running tests again with a slight tweak to remove some code that shouldn't be needed any more.

**Comment by wburns** [comment]
Tests still seem good rebasing.  Also changing exception to just use CacheException instead.

**Comment by tristantarrant** [comment]
Merged (after removing references to the CacheAlreadyDefinedException from the javadocs)

File: core/src/main/java/org/infinispan/manager/DefaultCacheManager.java [hunk 1]
```diff
@@ -349,26 +349,25 @@ public Configuration defineConfiguration(String name, String template, Configura
...
-            return defineConfiguration(name, false, configurationOverride, c);
+            return actualDefineConfiguration(name, configurationOverride, c);
          }
       }
-      return defineConfiguration(name, configurationOverride);
+      return actualDefineConfiguration(name, configurationOverride);
    }
 
-   private Configuration defineConfiguration(String name, boolean checkExisting, Configuration... configurations) {
+   private Configuration actualDefineConfiguration(String name, Configuration... configurations) {
```
Comment (danberindei) [inline core/src/main/java/org/infinispan/manager/DefaultCacheManager.java]: Nitpicking: I'd prefer `internalDefineConfiguration` or `doDefineConfiguration`, `actual` is an adjective so it fits better when delegating to another object.

File: core/src/main/java/org/infinispan/manager/DefaultCacheManager.java [hunk 2]
```diff
@@ -349,26 +349,25 @@ public Configuration defineConfiguration(String name, String template, Configura
          if (c == null) {
             throw log.undeclaredConfiguration(template, name);
          } else {
-            return defineConfiguration(name, false, configurationOverride, c);
+            return actualDefineConfiguration(name, configurationOverride, c);
```
Comment (danberindei) [inline core/src/main/java/org/infinispan/manager/DefaultCacheManager.java]: I'm not sure, but I think the `configurationOverride` should be last.

File: core/src/main/java/org/infinispan/manager/EmbeddedCacheManager.java [hunk 3]
```diff
@@ -239,6 +232,8 @@
     * @param configurationName name of the configuration template to use
     * @return null if no configuration exists as per rules set above, otherwise
     *         returns a cache instance identified by cacheName
+    * @deprecated Use {@link EmbeddedCacheManager#defineConfiguration(String, String, Configuration)} and
+    * {@link EmbeddedCacheManager#getCache(String)} instead
```
Comment (danberindei) [inline core/src/main/java/org/infinispan/manager/EmbeddedCacheManager.java]: I usually add a version number to make it easier to decide when to actually remove the deprecated classes/methods.

File: spring/spring4/spring4-embedded/src/test/java/org/infinispan/spring/support/embedded/InfinispanNamedEmbeddedCacheFactoryBeanTest.java [hunk 4]
```diff
@@ -42,8 +42,6 @@
    public void startCacheManagers() {
       DEFAULT_CACHE_MANAGER = TestCacheManagerFactory.createCacheManager();
       DEFAULT_CACHE_MANAGER.start();
-      DEFAULT_CACHE_MANAGER.defineConfiguration("test.cache.Name", DEFAULT_CACHE_MANAGER.getDefaultCacheConfiguration());
-      DEFAULT_CACHE_MANAGER.defineConfiguration("test.bean.Name", DEFAULT_CACHE_MANAGER.getDefaultCacheConfiguration());
```
Comment (danberindei) [inline spring/spring4/spring4-embedded/src/test/java/org/infinispan/spring/support/embedded/InfinispanNamedEmbeddedCacheFactoryBeanTest.java]: What was the problem here?

Comment (wburns) [inline spring/spring4/spring4-embedded/src/test/java/org/infinispan/spring/support/embedded/InfinispanNamedEmbeddedCacheFactoryBeanTest.java]: These were already defined elsewhere in the test (causing an exception) and defining it as default does nothing anyways.

File: core/src/main/java/org/infinispan/manager/DefaultCacheManager.java (hunk 2 above)
Comment (wburns) [inline core/src/main/java/org/infinispan/manager/DefaultCacheManager.java]: Hrmm maybe.  We can log a different JIRA for it me thinks, all I did was remove the boolean.

File: core/src/main/java/org/infinispan/manager/EmbeddedCacheManager.java (hunk 3 above)
Comment (wburns) [inline core/src/main/java/org/infinispan/manager/EmbeddedCacheManager.java]: Sure I can add that

File: core/src/main/java/org/infinispan/manager/DefaultCacheManager.java (hunk 1 above)
Comment (wburns) [inline core/src/main/java/org/infinispan/manager/DefaultCacheManager.java]: Sure I can change that along with the version

File: core/src/main/java/org/infinispan/manager/DefaultCacheManager.java (hunk 2 above)
Comment (danberindei) [inline core/src/main/java/org/infinispan/manager/DefaultCacheManager.java]: Yeah, I saw you didn't really change it, but I still assigned https://issues.jboss.org/browse/ISPN-7587 to you ;)

**Pull request closed without merging**